source/base/mpi.cc

```diff
@@ -812,7 +814,890 @@ namespace Utilities
+
+        void
+        reinit(const IndexSet &owned_indices, const MPI_Comm &comm)
+        {
+          this->partition(owned_indices, comm);
+
+#ifdef DEAL_II_WITH_MPI
+          unsigned int my_rank = this_mpi_process(comm);
+
+          types::global_dof_index              dic_local_rececived = 0;
+          std::map<unsigned int, unsigned int> relevant_procs_map;
+
+          // 2) collect relevant processes and process local dict entries
+          {
+            std::vector<unsigned int> relevant_procs;
+            for (auto i : owned_indices)
+              {
+                unsigned int other_rank = this->dof_to_dict_rank(i);
+                if (other_rank == my_rank)
+                  {
+                    this->actually_owning_ranks[i - this->local_range.first] =
+                      my_rank;
+                    dic_local_rececived++;
+                  }
+                else if (relevant_procs.empty() ||
+                         relevant_procs.back() != other_rank)
+                  relevant_procs.push_back(other_rank);
+              }
+
+            {
+              unsigned int c = 0;
+              for (auto i : relevant_procs)
+                relevant_procs_map[i] = c++;
+            }
+          }
+
+          const unsigned int n_relevant_procs = relevant_procs_map.size();
+          std::vector<std::vector<
+            std::pair<types::global_dof_index, types::global_dof_index>>>
+                                   buffers(n_relevant_procs);
+          std::vector<MPI_Request> request(n_relevant_procs);
+
+          // 3) send messages with local dofs to the right dict process
+          {
+            std::vector<std::vector<types::global_dof_index>> temp(
+              n_relevant_procs);
+
+            // collect dofs of each dict process
+            for (auto i : owned_indices)
+              {
+                unsigned int other_rank = this->dof_to_dict_rank(i);
+                if (other_rank != my_rank)
+                  temp[relevant_procs_map[other_rank]].push_back(i);
+              }
+
+            // send dofs to each process
+            for (auto rank_pair : relevant_procs_map)
+              {
+                int rank  = rank_pair.first;
+                int index = rank_pair.second;
+
+                // create index set and compress data to be sent
+                auto &   indices_i = temp[index];
+                IndexSet is(this->size);
+                is.add_indices(indices_i.begin(), indices_i.end());
+                is.compress();
+
+                // translate index set to a list of pairs
+                auto &buffer = buffers[index];
+                for (auto interval = is.begin_intervals();
+                     interval != is.end_intervals();
+                     interval++)
+                  buffer.emplace_back(*interval->begin(), interval->last() + 1);
+
+                // send data
+                auto ierr = MPI_Isend(buffer.data(),
```

Comment:
```suggestion
                const auto ierr = MPI_Isend(buffer.data(),
```
and elsewhere if it is not reused.